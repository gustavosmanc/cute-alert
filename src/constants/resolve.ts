export const ALERT_RESOLVE_VALUES = {
  PRIMARY_BUTTON_CLICKED: 'primaryButtonClicked',
  SECONDARY_BUTTON_CLICKED: 'secondaryButtonClicked',
  CLOSE: 'close'
} as const

export const TOAST_RESOLVE_VALUES = {
  CLOSE: 'close'
} as const
