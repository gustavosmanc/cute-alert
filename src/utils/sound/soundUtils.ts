export const playSound = (src?: string) => {
  if (src) {
    const sound = new Audio(src)
    sound.play()
  }
}
