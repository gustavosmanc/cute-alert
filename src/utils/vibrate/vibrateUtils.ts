export const vibrate = (pattern: number | number[]) => {
  navigator.vibrate(pattern)
}
