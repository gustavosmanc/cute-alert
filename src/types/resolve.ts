import { ALERT_RESOLVE_VALUES, TOAST_RESOLVE_VALUES } from '@/constants/resolve'

export type AlertResolveValue = (typeof ALERT_RESOLVE_VALUES)[keyof typeof ALERT_RESOLVE_VALUES]
export type ToastResolveValue = (typeof TOAST_RESOLVE_VALUES)[keyof typeof TOAST_RESOLVE_VALUES]

export type AlertResolve = (value: AlertResolveValue) => void
export type ToastResolve = (value: ToastResolveValue) => void
