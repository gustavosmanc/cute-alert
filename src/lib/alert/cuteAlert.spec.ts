import { ALERT_RESOLVE_VALUES } from '@/constants/resolve'
import * as soundUtils from '@/utils/sound/soundUtils'
import * as vibrateUtils from '@/utils/vibrate/vibrateUtils'
import { queryByText } from '@testing-library/dom'
import '@testing-library/jest-dom'
import { cuteAlert } from '..'

const vibrateSpy = vi.spyOn(vibrateUtils, 'vibrate').mockImplementation(() => {})
const playSoundSpy = vi.spyOn(soundUtils, 'playSound').mockImplementation(() => {})

const getExampleDOM = () => {
  document.body = document.createElement('body')

  return document.body
}

const originalGlobalURL = global.URL

describe('cuteAlert', () => {
  describe('snapshots', () => {
    beforeEach(() => {
      // @ts-expect-error
      global.URL = class URL {
        constructor(url: string) {
          // @ts-expect-error
          this.url = url
        }

        get href() {
          // @ts-expect-error
          return this.url
        }
      }
    })

    afterEach(() => {
      global.URL = originalGlobalURL
    })

    it('should render success alert', () => {
      const container = getExampleDOM()

      cuteAlert({
        type: 'success',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render error alert', () => {
      const container = getExampleDOM()

      cuteAlert({
        type: 'error',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render info alert', () => {
      const container = getExampleDOM()

      cuteAlert({
        type: 'info',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render warning alert', () => {
      const container = getExampleDOM()

      cuteAlert({
        type: 'warning',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render question alert', () => {
      const container = getExampleDOM()

      cuteAlert({
        type: 'question',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render alert without title', () => {
      const container = getExampleDOM()

      cuteAlert({
        type: 'success',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render alert without description', () => {
      const container = getExampleDOM()

      cuteAlert({
        type: 'success',
        title: 'Title',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render alert without primary button', () => {
      const container = getExampleDOM()

      cuteAlert({
        type: 'success',
        title: 'Title',
        description: 'Description',
        secondaryButtonText: 'Cancel'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render alert without secondary button', () => {
      const container = getExampleDOM()

      cuteAlert({
        type: 'success',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render alert with timer', () => {
      const container = getExampleDOM()

      cuteAlert({
        type: 'success',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel',
        timer: 5000
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render alert with circle close style', () => {
      const container = getExampleDOM()

      cuteAlert({
        type: 'success',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel',
        closeStyle: 'circle'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render alert with custom image size', () => {
      const container = getExampleDOM()

      cuteAlert({
        type: 'success',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel',
        imageSize: 48
      })

      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should render alert with custom image source', () => {
      const container = getExampleDOM()

      cuteAlert({
        type: 'success',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel',
        imageSrc: 'random image source'
      })

      expect(container.innerHTML).toMatchSnapshot()
    })
  })

  describe('functionalities', () => {
    it('should handle vibrate and playSound when alert is open', () => {
      getExampleDOM()

      cuteAlert({
        type: 'success',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel',
        vibrationPattern: [1, 2, 3],
        soundSrc: 'sound source'
      })

      expect(vibrateSpy).toHaveBeenCalledTimes(1)
      expect(vibrateSpy).toHaveBeenCalledWith([1, 2, 3])
      expect(playSoundSpy).toHaveBeenCalledTimes(1)
      expect(playSoundSpy).toHaveBeenCalledWith('sound source')
    })

    it('should close previous alert when opening a new one', () => {
      const container = getExampleDOM()

      cuteAlert({
        type: 'success',
        title: 'First Alert Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel'
      })

      cuteAlert({
        type: 'success',
        title: 'Second Alert Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel'
      })

      expect(queryByText(container, 'First Alert Title')).not.toBeInTheDocument()
      expect(queryByText(container, 'Second Alert Title')).toBeInTheDocument()
    })

    it('should handle primary button click', async () => {
      const callbackMock = vi.fn()
      const container = getExampleDOM()

      cuteAlert({
        type: 'success',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel'
      }).then((response) => callbackMock(response))

      queryByText(container, 'Confirm').click()

      await new Promise(process.nextTick)

      expect(container).toBeEmptyDOMElement()
      expect(callbackMock).toHaveBeenCalledTimes(1)
      expect(callbackMock).toHaveBeenCalledWith(ALERT_RESOLVE_VALUES.PRIMARY_BUTTON_CLICKED)
    })

    it('should handle secondary button click', async () => {
      const callbackMock = vi.fn()
      const container = getExampleDOM()

      cuteAlert({
        type: 'success',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel'
      }).then((response) => callbackMock(response))

      queryByText(container, 'Cancel').click()

      await new Promise(process.nextTick)

      expect(container).toBeEmptyDOMElement()
      expect(callbackMock).toHaveBeenCalledTimes(1)
      expect(callbackMock).toHaveBeenCalledWith(ALERT_RESOLVE_VALUES.SECONDARY_BUTTON_CLICKED)
    })

    it('should handle close button click', async () => {
      const callbackMock = vi.fn()
      const container = getExampleDOM()

      cuteAlert({
        type: 'success',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel'
      }).then((response) => callbackMock(response))

      queryByText(container, 'X').click()

      await new Promise(process.nextTick)

      expect(container).toBeEmptyDOMElement()
      expect(callbackMock).toHaveBeenCalledTimes(1)
      expect(callbackMock).toHaveBeenCalledWith(ALERT_RESOLVE_VALUES.CLOSE)
    })

    it('should handle close by timer', async () => {
      vi.useFakeTimers()

      const callbackMock = vi.fn()
      const container = getExampleDOM()

      cuteAlert({
        type: 'success',
        title: 'Title',
        description: 'Description',
        primaryButtonText: 'Confirm',
        secondaryButtonText: 'Cancel',
        timer: 5000
      }).then((response) => callbackMock(response))

      vi.advanceTimersByTime(4999)
      await new Promise(process.nextTick)

      expect(container).not.toBeEmptyDOMElement()
      expect(callbackMock).not.toHaveBeenCalled()

      vi.advanceTimersByTime(1)
      await new Promise(process.nextTick)

      expect(container).toBeEmptyDOMElement()
      expect(callbackMock).toHaveBeenCalledTimes(1)
      expect(callbackMock).toHaveBeenCalledWith(ALERT_RESOLVE_VALUES.CLOSE)

      vi.useRealTimers()
    })
  })
})
