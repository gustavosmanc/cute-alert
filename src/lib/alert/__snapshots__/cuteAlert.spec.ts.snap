// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cuteAlert > snapshots > should render alert with circle close style 1`] = `
"<div class="cute-alert cute-alert--success">
      <div class="cute-alert__frame">
        <div class="cute-alert__header">
          <span class="cute-alert__close cute-alert__close--circle">X</span>
          <img class="cute-alert__image" src="/src/lib/icons/success.svg" height="80" width="80">
        </div>
        <div class="cute-alert__body">
          <span class="cute-alert__title">Title</span>
          <span class="cute-alert__description">Description</span>
          <div class="cute-alert__buttons">
            <button class="cute-alert__primary-btn">Confirm</button>
            <button class="cute-alert__secondary-btn">Cancel</button>
          </div>
          
        </div>
      </div>
    </div>"
`;

exports[`cuteAlert > snapshots > should render alert with custom image size 1`] = `
"<div class="cute-alert cute-alert--success">
      <div class="cute-alert__frame">
        <div class="cute-alert__header">
          <span class="cute-alert__close cute-alert__close--default">X</span>
          <img class="cute-alert__image" src="/src/lib/icons/success.svg" height="48" width="48">
        </div>
        <div class="cute-alert__body">
          <span class="cute-alert__title">Title</span>
          <span class="cute-alert__description">Description</span>
          <div class="cute-alert__buttons">
            <button class="cute-alert__primary-btn">Confirm</button>
            <button class="cute-alert__secondary-btn">Cancel</button>
          </div>
          
        </div>
      </div>
    </div>"
`;

exports[`cuteAlert > snapshots > should render alert with custom image source 1`] = `
"<div class="cute-alert cute-alert--success">
      <div class="cute-alert__frame">
        <div class="cute-alert__header">
          <span class="cute-alert__close cute-alert__close--default">X</span>
          <img class="cute-alert__image" src="random image source" height="80" width="80">
        </div>
        <div class="cute-alert__body">
          <span class="cute-alert__title">Title</span>
          <span class="cute-alert__description">Description</span>
          <div class="cute-alert__buttons">
            <button class="cute-alert__primary-btn">Confirm</button>
            <button class="cute-alert__secondary-btn">Cancel</button>
          </div>
          
        </div>
      </div>
    </div>"
`;

exports[`cuteAlert > snapshots > should render alert with timer 1`] = `
"<div class="cute-alert cute-alert--success">
      <div class="cute-alert__frame">
        <div class="cute-alert__header">
          <span class="cute-alert__close cute-alert__close--default">X</span>
          <img class="cute-alert__image" src="/src/lib/icons/success.svg" height="80" width="80">
        </div>
        <div class="cute-alert__body">
          <span class="cute-alert__title">Title</span>
          <span class="cute-alert__description">Description</span>
          <div class="cute-alert__buttons">
            <button class="cute-alert__primary-btn">Confirm</button>
            <button class="cute-alert__secondary-btn">Cancel</button>
          </div>
          <div class="cute-alert__timer" style="animation: timer 5000ms linear"></div>
        </div>
      </div>
    </div>"
`;

exports[`cuteAlert > snapshots > should render alert without description 1`] = `
"<div class="cute-alert cute-alert--success">
      <div class="cute-alert__frame">
        <div class="cute-alert__header">
          <span class="cute-alert__close cute-alert__close--default">X</span>
          <img class="cute-alert__image" src="/src/lib/icons/success.svg" height="80" width="80">
        </div>
        <div class="cute-alert__body">
          <span class="cute-alert__title">Title</span>
          
          <div class="cute-alert__buttons">
            <button class="cute-alert__primary-btn">Confirm</button>
            <button class="cute-alert__secondary-btn">Cancel</button>
          </div>
          
        </div>
      </div>
    </div>"
`;

exports[`cuteAlert > snapshots > should render alert without primary button 1`] = `
"<div class="cute-alert cute-alert--success">
      <div class="cute-alert__frame">
        <div class="cute-alert__header">
          <span class="cute-alert__close cute-alert__close--default">X</span>
          <img class="cute-alert__image" src="/src/lib/icons/success.svg" height="80" width="80">
        </div>
        <div class="cute-alert__body">
          <span class="cute-alert__title">Title</span>
          <span class="cute-alert__description">Description</span>
          <div class="cute-alert__buttons">
            
            <button class="cute-alert__secondary-btn">Cancel</button>
          </div>
          
        </div>
      </div>
    </div>"
`;

exports[`cuteAlert > snapshots > should render alert without secondary button 1`] = `
"<div class="cute-alert cute-alert--success">
      <div class="cute-alert__frame">
        <div class="cute-alert__header">
          <span class="cute-alert__close cute-alert__close--default">X</span>
          <img class="cute-alert__image" src="/src/lib/icons/success.svg" height="80" width="80">
        </div>
        <div class="cute-alert__body">
          <span class="cute-alert__title">Title</span>
          <span class="cute-alert__description">Description</span>
          <div class="cute-alert__buttons">
            <button class="cute-alert__primary-btn">Confirm</button>
            
          </div>
          
        </div>
      </div>
    </div>"
`;

exports[`cuteAlert > snapshots > should render alert without title 1`] = `
"<div class="cute-alert cute-alert--success">
      <div class="cute-alert__frame">
        <div class="cute-alert__header">
          <span class="cute-alert__close cute-alert__close--default">X</span>
          <img class="cute-alert__image" src="/src/lib/icons/success.svg" height="80" width="80">
        </div>
        <div class="cute-alert__body">
          
          <span class="cute-alert__description">Description</span>
          <div class="cute-alert__buttons">
            <button class="cute-alert__primary-btn">Confirm</button>
            <button class="cute-alert__secondary-btn">Cancel</button>
          </div>
          
        </div>
      </div>
    </div>"
`;

exports[`cuteAlert > snapshots > should render error alert 1`] = `
"<div class="cute-alert cute-alert--error">
      <div class="cute-alert__frame">
        <div class="cute-alert__header">
          <span class="cute-alert__close cute-alert__close--default">X</span>
          <img class="cute-alert__image" src="/src/lib/icons/error.svg" height="80" width="80">
        </div>
        <div class="cute-alert__body">
          <span class="cute-alert__title">Title</span>
          <span class="cute-alert__description">Description</span>
          <div class="cute-alert__buttons">
            <button class="cute-alert__primary-btn">Confirm</button>
            <button class="cute-alert__secondary-btn">Cancel</button>
          </div>
          
        </div>
      </div>
    </div>"
`;

exports[`cuteAlert > snapshots > should render info alert 1`] = `
"<div class="cute-alert cute-alert--info">
      <div class="cute-alert__frame">
        <div class="cute-alert__header">
          <span class="cute-alert__close cute-alert__close--default">X</span>
          <img class="cute-alert__image" src="/src/lib/icons/info.svg" height="80" width="80">
        </div>
        <div class="cute-alert__body">
          <span class="cute-alert__title">Title</span>
          <span class="cute-alert__description">Description</span>
          <div class="cute-alert__buttons">
            <button class="cute-alert__primary-btn">Confirm</button>
            <button class="cute-alert__secondary-btn">Cancel</button>
          </div>
          
        </div>
      </div>
    </div>"
`;

exports[`cuteAlert > snapshots > should render question alert 1`] = `
"<div class="cute-alert cute-alert--question">
      <div class="cute-alert__frame">
        <div class="cute-alert__header">
          <span class="cute-alert__close cute-alert__close--default">X</span>
          <img class="cute-alert__image" src="/src/lib/icons/question.svg" height="80" width="80">
        </div>
        <div class="cute-alert__body">
          <span class="cute-alert__title">Title</span>
          <span class="cute-alert__description">Description</span>
          <div class="cute-alert__buttons">
            <button class="cute-alert__primary-btn">Confirm</button>
            <button class="cute-alert__secondary-btn">Cancel</button>
          </div>
          
        </div>
      </div>
    </div>"
`;

exports[`cuteAlert > snapshots > should render success alert 1`] = `
"<div class="cute-alert cute-alert--success">
      <div class="cute-alert__frame">
        <div class="cute-alert__header">
          <span class="cute-alert__close cute-alert__close--default">X</span>
          <img class="cute-alert__image" src="/src/lib/icons/success.svg" height="80" width="80">
        </div>
        <div class="cute-alert__body">
          <span class="cute-alert__title">Title</span>
          <span class="cute-alert__description">Description</span>
          <div class="cute-alert__buttons">
            <button class="cute-alert__primary-btn">Confirm</button>
            <button class="cute-alert__secondary-btn">Cancel</button>
          </div>
          
        </div>
      </div>
    </div>"
`;

exports[`cuteAlert > snapshots > should render warning alert 1`] = `
"<div class="cute-alert cute-alert--warning">
      <div class="cute-alert__frame">
        <div class="cute-alert__header">
          <span class="cute-alert__close cute-alert__close--default">X</span>
          <img class="cute-alert__image" src="/src/lib/icons/warning.svg" height="80" width="80">
        </div>
        <div class="cute-alert__body">
          <span class="cute-alert__title">Title</span>
          <span class="cute-alert__description">Description</span>
          <div class="cute-alert__buttons">
            <button class="cute-alert__primary-btn">Confirm</button>
            <button class="cute-alert__secondary-btn">Cancel</button>
          </div>
          
        </div>
      </div>
    </div>"
`;
